### PR TITLE
Improve "loading in progress" detection and solve bug #1605

### PR DESCRIPTION
**Pull request by freddidierRTE**
in release note : 

Bug:

 On startup of the application, sometimes loading in progress appears when no cards are available for user (#1605)

### CODE REVIEW

File: ui/main/src/app/services/lightcards.service.ts [hunk 1]
```diff
@@ -92,24 +97,39 @@ export class LightCardsService {
...
+        if (this.timeOfLastCardReception > this.timeOfLastDebounce) // at least one card has been received but not yet process by getLightCardDebounce() , it is in progress
+                  {
+                    this.numberOfFollowingTimeDebounceHasNotBeenDone++;
+                    if (this.numberOfFollowingTimeDebounceHasNotBeenDone > 2) this.loadingInProgress.next(true);
+                  }
+        else  {
+            this.loadingInProgress.next(false);
+            this.numberOfFollowingTimeDebounceHasNotBeenDone = 0;
+        }
+        setTimeout(() => this.checkForLoadingInProgress(), 500);
```
Comment (AlexGuironnetRTE) [inline ui/main/src/app/services/lightcards.service.ts:130]: Defining the various milliseconds durations or card numbers as constants might also help?

**Review by AlexGuironnetRTE — CHANGES_REQUESTED** [review]
Tests with new version OK. No error message, no visible change to performance.
I haven't been able to reproduce the issue with the old code, but in any case, checking the "loading in progress" status with a timeout as it is now done should prevent it getting "stuck" as described in the issue.

I have a few suggestions on the naming and comments (see comments) but other than that it's ok for me.

File: ui/main/src/app/services/lightcards.service.ts [hunk 2]
```diff
@@ -22,19 +22,24 @@ import {LightCard, Severity} from '@ofModel/light-card.model';
 export class LightCardsService {
 
     private filters: Filter[];
-    private lastDebounce: number = 0;
+
+    private timeOfLastDebounce: number = 0;
+    private timeOfLastCardReception: number = 0;
+    private numberOfFollowingTimeDebounceHasNotBeenDone: number=0;
```
[the remark this answers is not in the document]
Comment (freddidierRTE) [reply ui/main/src/app/services/lightcards.service.ts]: Done

File: ui/main/src/app/services/lightcards.service.ts [hunk 3]
```diff
@@ -92,24 +97,39 @@ export class LightCardsService {
...
+            tap(() => this.timeOfLastDebounce = (new Date()).valueOf())
         );
     }
 
     public getLoadingInProgress() {
         return this.loadingInProgress.asObservable();
     }
 
+    // check every 500ms if we are in a loading card process
+    // we consider we are in a loading card process if we have 3 following times cards not processed by getLightCardDebounce();
```
[the remark this answers is not in the document]
Comment (freddidierRTE) [reply ui/main/src/app/services/lightcards.service.ts]: Done

File: ui/main/src/app/services/lightcards.service.ts [hunk 4]
```diff
@@ -92,24 +97,39 @@ export class LightCardsService {
...
+    private checkForLoadingInProgress()
+    {
+        if (this.timeOfLastCardReception > this.timeOfLastDebounce) // at least one card has been received but not yet process by getLightCardDebounce() , it is in progress
```
[the remark this answers is not in the document]
Comment (freddidierRTE) [reply ui/main/src/app/services/lightcards.service.ts]: Done

File: ui/main/src/app/services/lightcards.service.ts (hunk 1 above)
Comment (freddidierRTE) [reply ui/main/src/app/services/lightcards.service.ts]: yes on principle but in this case , i think it is obvious

**Pull request merged**